File: scripts/import_whale_composite.py
#!/usr/bin/env python3
"""
Import Whale Composite Data to Supabase

Imports the FY27 Q1 whale accounts from the merged Demandbase + ZoomInfo CSV.
"""

import pandas as pd
import requests
import json
import sys
from pathlib import Path

# =============================================================================
# CONFIG
# =============================================================================

CSV_PATH = "/Users/arijitchowdhury/Library/CloudStorage/GoogleDrive-arijit.chowdhury@algolia.com/My Drive/AI/DATA/2026-02-27-Whale_Demandbase+ZoomInfo.csv"
SUPABASE_URL = "https://xbitqeejsgqnwvxlnjra.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InhiaXRxZWVqc2dxbnd2eGxuanJhIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzIwODU1NDAsImV4cCI6MjA4NzY2MTU0MH0.XoEOx8rHo_1EyCF4yJ3g2S3tXUX_XepQu9PSfUWvyIg"
SOURCE_FILE = "2026-02-27-Whale_Demandbase+ZoomInfo.csv"
BATCH_SIZE = 50

# =============================================================================
# COLUMN MAPPING
# =============================================================================

COLUMN_MAP = {
    # Core identifiers
    'Domain': 'domain',
    'Account Name': 'account_name',
    '\ufeffAccount Name': 'account_name',  # BOM variant
    '18 digit Account ID': 'salesforce_account_id',
    'ABM ID': 'abm_id',
    'acc_id': 'acc_id',
    'ZoomInfo Company ID': 'zoominfo_company_id',

    # Demandbase ABX
    'Journey Stage': 'journey_stage',
    'Days in Journey Stage': 'days_in_journey_stage',
    'Date of Last Web Activity': 'date_of_last_web_activity',
    'Engaged Known People': 'engaged_known_people',
    'Engagement Points (3 mo.)': 'engagement_points_3mo',
    'Target Account': 'target_account',
    'ABX Status': 'abx_status',
    'ABX Status Reason': 'abx_status_reason',
    'ABX Status Reason Description': 'abx_status_reason_description',
    'GTM Tag': 'gtm_tag',

    # Demandbase firmographics
    'Billing Country': 'billing_country',
    'NAICS Code': 'naics_code',
    'NAICS Code Priority 1': 'naics_code_priority_1',
    'NAICS Description Priority 1': 'naics_description_priority_1',
    'Industry': 'industry',
    'Demandbase Industry': 'demandbase_industry',
    'Demandbase Sub Industry': 'demandbase_sub_industry',
    'Revenue': 'revenue',
    'Revenue Range': 'revenue_range',
    'Expected Revenue': 'expected_revenue',
    'ARR': 'arr',
    'Traffic': 'traffic',

    # Sales assignment
    'Local Segment': 'local_segment',
    'Account Region': 'account_region',
    'Demandbase - Account Owner Name': 'demandbase_account_owner',
    'Sales Region': 'sales_region',
    'Sales Sub-Region': 'sales_sub_region',
    'Sales Segment': 'sales_segment',

    # Technology flags
    'Bigcommerce Technology': 'has_bigcommerce',
    'Commercetools Technology': 'has_commercetools',
    'Magento Technology': 'has_magento',
    'Magento Open Source Technology': 'has_magento_open_source',
    'Salesforce B2b Commerce Technology': 'has_salesforce_b2b_commerce',
    'Salesforce Commerce Cloud Technology': 'has_salesforce_commerce_cloud',
    'Shopify Hosted Technology': 'has_shopify_hosted',
    'Shopify Plus Technology': 'has_shopify_plus',
    'Shopify Technology': 'has_shopify',
    'Spryker Technology': 'has_spryker',

    # ZoomInfo company data
    'Company Name': 'zi_company_name',
    'Company Description': 'zi_company_description',
    'Website': 'zi_website',
    'Founded Year': 'zi_founded_year',
    'Company HQ Phone': 'zi_company_hq_phone',
    'Fax': 'zi_fax',
    'Ticker': 'zi_ticker',

    # ZoomInfo financials
    'Revenue (in 000s USD)': 'zi_revenue_thousands',
    'Revenue Range (in USD)': 'zi_revenue_range',
    'Est. Marketing Department Budget (in 000s USD)': 'zi_est_marketing_budget_thousands',
    'Est. Finance Department Budget (in 000s USD)': 'zi_est_finance_budget_thousands',
    'Est. IT Department Budget (in 000s USD)': 'zi_est_it_budget_thousands',
    'Est. HR Department Budget (in 000s USD)': 'zi_est_hr_budget_thousands',

    # ZoomInfo employees
    'Employees': 'zi_employees',
    'Employee Range': 'zi_employee_range',
    'Past 1 Year Employee Growth Rate': 'zi_employee_growth_1yr',
    'Past 2 Year Employee Growth Rate': 'zi_employee_growth_2yr',

    # ZoomInfo industry
    'SIC Code 1': 'zi_sic_code_1',
    'SIC Code 2': 'zi_sic_code_2',
    'SIC Codes': 'zi_sic_codes',
    'NAICS Code 1': 'zi_naics_code_1',
    'NAICS Code 2': 'zi_naics_code_2',
    'NAICS Codes': 'zi_naics_codes',
    'Primary Industry': 'zi_primary_industry',
    'Primary Sub-Industry': 'zi_primary_sub_industry',
    'All Industries': 'zi_all_industries',
    'All Sub-Industries': 'zi_all_sub_industries',
    'Industry Hierarchical Category': 'zi_industry_hierarchical_category',
    'Secondary Industry Hierarchical Category': 'zi_secondary_industry_hierarchical_category',

    # ZoomInfo digital presence
    'Alexa Rank': 'zi_alexa_rank',
    'ZoomInfo Company Profile URL': 'zi_profile_url',
    'LinkedIn Company Profile URL': 'zi_linkedin_url',
    'Facebook Company Profile URL': 'zi_facebook_url',
    'Twitter Company Profile URL': 'zi_twitter_url',

    # ZoomInfo corporate
    'Ownership Type': 'zi_ownership_type',
    'Business Model': 'zi_business_model',
    'Certified Active Company': 'zi_certified_active',
    'Certification Date': 'zi_certification_date',
    'Defunct Company': 'zi_defunct_company',
    'Company Is Acquired': 'zi_company_is_acquired',
    'Number of Locations': 'zi_number_of_locations',

    # ZoomInfo funding
    'Total Funding Amount (in 000s USD)': 'zi_total_funding_thousands',
    'Recent Funding Amount (in 000s USD)': 'zi_recent_funding_thousands',
    'Recent Funding Round': 'zi_recent_funding_round',
    'Recent Funding Date': 'zi_recent_funding_date',
    'Recent Investors': 'zi_recent_investors',
    'All Investors': 'zi_all_investors',

    # ZoomInfo address
    'Company Street Address': 'zi_street_address',
    'Company City': 'zi_city',
    'Company State': 'zi_state',
    'Company Zip Code': 'zi_zip_code',
    'Company Country': 'zi_country',
    'Full Address': 'zi_full_address',

    # ZoomInfo corporate hierarchy
    'Company ID (Ultimate Parent)': 'zi_ultimate_parent_id',
    'Entity Name (Ultimate Parent)': 'zi_ultimate_parent_name',
    'Company ID (Immediate Parent)': 'zi_immediate_parent_id',
    'Entity Name (Immediate Parent)': 'zi_immediate_parent_name',
    'Relationship (Immediate Parent)': 'zi_immediate_parent_relationship',

    # ZoomInfo match quality
    'Match status': 'zi_match_status',
    'Company Name Match Insight': 'zi_match_insight_name',
    'Company Domain Match Insight': 'zi_match_insight_domain',
    'Company Address Match Insight': 'zi_match_insight_address',
    'Company Phone Match Insight': 'zi_match_insight_phone',
    'Company ID Match Insight': 'zi_match_insight_id',
    'Company Social URL Match Insight': 'zi_match_insight_social',
}

# =============================================================================
# HELPERS
# =============================================================================

def clean_value(val):
    """Clean a value for JSON serialization."""
    if pd.isna(val) or val == '' or val == 'null':
        return None
    if isinstance(val, float) and val != val:  # NaN check
        return None
    return val

def clean_domain(domain):
    """Clean and normalize domain."""
    if not domain or pd.isna(domain):
        return None
    domain = str(domain).strip().lower()
    domain = domain.replace('www.', '').rstrip('/')
    if '.' not in domain:
        return None
    return domain

def to_bool(val):
    """Convert to boolean."""
    if pd.isna(val):
        return False
    return str(val).lower() == 'true'

def to_int(val):
    """Convert to integer."""
    if pd.isna(val) or val == '':
        return None
    try:
        return int(float(str(val).replace(',', '')))
    except (ValueError, TypeError):
        return None

def to_float(val):
    """Convert to float."""
    if pd.isna(val) or val == '':
        return None
    try:
        return float(str(val).replace(',', ''))
    except (ValueError, TypeError):
        return None

def map_row(row):
    """Map a DataFrame row to database columns."""
    record = {}

    for csv_col, db_col in COLUMN_MAP.items():
        if csv_col not in row:
            continue

        val = row[csv_col]

        # Handle boolean columns
        if db_col.startswith('has_'):
            record[db_col] = to_bool(val)
        # Handle integer columns
        elif db_col in ['days_in_journey_stage', 'engaged_known_people', 'acc_id',
                        'zi_founded_year', 'zi_employees', 'zi_alexa_rank',
                        'zi_number_of_locations', 'revenue', 'traffic',
                        'zi_revenue_thousands', 'zi_est_marketing_budget_thousands',
                        'zi_est_finance_budget_thousands', 'zi_est_it_budget_thousands',
                        'zi_est_hr_budget_thousands', 'zi_total_funding_thousands',
                        'zi_recent_funding_thousands', 'zoominfo_company_id']:
            record[db_col] = to_int(val)
        # Handle float columns
        elif db_col in ['engagement_points_3mo', 'expected_revenue', 'arr',
                        'zi_employee_growth_1yr', 'zi_employee_growth_2yr']:
            record[db_col] = to_float(val)
        # Handle domain specially
        elif db_col == 'domain':
            record[db_col] = clean_domain(val)
        else:
            record[db_col] = clean_value(val)

    # Add source file
    record['source_file'] = SOURCE_FILE

    return record

# =============================================================================
# MAIN
# =============================================================================

def main():
    print("=" * 60)
    print("WHALE COMPOSITE IMPORT")
    print("=" * 60)

    # Read CSV
    print(f"\nReading CSV: {CSV_PATH}")
    df = pd.read_csv(CSV_PATH, dtype=str)  # Read all as string first
    print(f"Loaded {len(df)} rows, {len(df.columns)} columns")

    # Map rows
    print("\nMapping rows to database schema...")
    records = []
    skipped = 0

    for idx, row in df.iterrows():
        mapped = map_row(row)
        if mapped.get('domain'):
            records.append(mapped)
        else:
            skipped += 1

    print(f"Mapped {len(records)} valid records, skipped {skipped}")

    # Deduplicate by domain (keep last occurrence - usually more complete)
    seen_domains = {}
    for record in records:
        domain = record.get('domain')
        if domain:
            seen_domains[domain] = record

    records = list(seen_domains.values())
    print(f"After deduplication: {len(records)} unique domains")

    # Insert in batches
    print(f"\nInserting to Supabase in batches of {BATCH_SIZE}...")

    headers = {
        'apikey': SUPABASE_KEY,
        'Authorization': f'Bearer {SUPABASE_KEY}',
        'Content-Type': 'application/json',
        'Prefer': 'resolution=merge-duplicates'  # Upsert on conflict
    }

    # Use on_conflict to enable proper upsert
    url = f"{SUPABASE_URL}/rest/v1/whale_composite?on_conflict=domain"
    inserted = 0
    errors = 0

    for i in range(0, len(records), BATCH_SIZE):
        batch = records[i:i + BATCH_SIZE]

        # Convert to JSON, handling any remaining issues
        try:
            payload = json.dumps(batch, default=str)
        except Exception as e:
            print(f"\nJSON error at batch {i // BATCH_SIZE + 1}: {e}")
            errors += len(batch)
            continue

        response = requests.post(url, headers=headers, data=payload)

        if response.status_code in [200, 201]:
            inserted += len(batch)
            sys.stdout.write(f"\rInserted: {inserted}/{len(records)}")
            sys.stdout.flush()
        else:
            print(f"\nBatch {i // BATCH_SIZE + 1} error: {response.status_code}")
            print(response.text[:200])
            errors += len(batch)

    print("\n")
    print("=" * 60)
    print("IMPORT COMPLETE")
    print("=" * 60)
    print(f"Total records: {len(records)}")
    print(f"Inserted: {inserted}")
    print(f"Errors: {errors}")

    # Verify
    print("\nVerifying...")
    verify_url = f"{SUPABASE_URL}/rest/v1/whale_composite?select=count"
    verify_headers = {
        'apikey': SUPABASE_KEY,
        'Authorization': f'Bearer {SUPABASE_KEY}',
        'Prefer': 'count=exact'
    }
    verify_response = requests.get(f"{SUPABASE_URL}/rest/v1/whale_composite?select=id", headers=verify_headers)

    if 'content-range' in verify_response.headers:
        count = verify_response.headers['content-range'].split('/')[1]
        print(f"Records in database: {count}")
    else:
        print(f"Records in response: {len(verify_response.json())}")

if __name__ == '__main__':
    main()

File: scripts/import-whale-composite.ts
/**
 * Import Whale Composite Data
 *
 * Imports the FY27 Q1 whale accounts from the merged Demandbase + ZoomInfo CSV
 * into the whale_composite Supabase table.
 *
 * Usage:
 *   npx ts-node scripts/import-whale-composite.ts
 *
 * Requires:
 *   - SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY in environment
 *   - The whale_composite table to exist (run migration first)
 */

import * as fs from 'fs';
import * as path from 'path';
import { parse } from 'csv-parse/sync';
import { createClient } from '@supabase/supabase-js';

// =============================================================================
// CONFIG
// =============================================================================

const CSV_PATH = path.join(__dirname, '../../../AI/DATA/2026-02-27_FY27-Q1-Whale_Demandbase+ZoomInfo.csv');
const SOURCE_FILE = '2026-02-27_FY27-Q1-Whale_Demandbase+ZoomInfo.csv';

// Load from environment or use defaults for local dev
const SUPABASE_URL = process.env.SUPABASE_URL || 'https://xbitqeejsgqnwvxlnjra.supabase.co';
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!SUPABASE_SERVICE_KEY) {
  console.error('ERROR: SUPABASE_SERVICE_ROLE_KEY environment variable required');
  console.error('Run: export SUPABASE_SERVICE_ROLE_KEY=your_key');
  process.exit(1);
}

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY);

// =============================================================================
// COLUMN MAPPING: CSV Header -> DB Column
// =============================================================================

function mapRow(row: Record<string, string>): Record<string, unknown> {
  // Helper to parse numbers
  const toInt = (v: string | undefined): number | null => {
    if (!v || v === '') return null;
    const n = parseInt(v.replace(/,/g, ''), 10);
    return isNaN(n) ? null : n;
  };

  const toFloat = (v: string | undefined): number | null => {
    if (!v || v === '') return null;
    const n = parseFloat(v.replace(/,/g, ''));
    return isNaN(n) ? null : n;
  };

  const toBool = (v: string | undefined): boolean => {
    return v?.toLowerCase() === 'true';
  };

  const toStr = (v: string | undefined): string | null => {
    if (!v || v === '' || v === 'null') return null;
    return v.trim();
  };

  // Extract domain, clean it
  let domain = toStr(row['Domain']);
  if (domain) {
    domain = domain.toLowerCase().replace(/^www\./, '').replace(/\/$/, '');
  }

  if (!domain) {
    console.warn('Skipping row with no domain:', row['Account Name'] || row['﻿Account Name']);
    return {};
  }

  return {
    // Core identifiers
    domain,
    account_name: toStr(row['Account Name'] || row['﻿Account Name']),
    salesforce_account_id: toStr(row['18 digit Account ID']),
    abm_id: toStr(row['ABM ID']),
    acc_id: toInt(row['acc_id']),
    zoominfo_company_id: toInt(row['ZoomInfo Company ID']),

    // Demandbase ABX
    journey_stage: toStr(row['Journey Stage']),
    days_in_journey_stage: toInt(row['Days in Journey Stage']),
    date_of_last_web_activity: toStr(row['Date of Last Web Activity']),
    engaged_known_people: toInt(row['Engaged Known People']),
    engagement_points_3mo: toFloat(row['Engagement Points (3 mo.)']),
    target_account: toStr(row['Target Account']),
    abx_status: toStr(row['ABX Status']),
    abx_status_reason: toStr(row['ABX Status Reason']),
    abx_status_reason_description: toStr(row['ABX Status Reason Description']),
    gtm_tag: toStr(row['GTM Tag']),

    // Demandbase firmographics
    billing_country: toStr(row['Billing Country']),
    naics_code: toStr(row['NAICS Code']),
    naics_code_priority_1: toStr(row['NAICS Code Priority 1']),
    naics_description_priority_1: toStr(row['NAICS Description Priority 1']),
    industry: toStr(row['Industry']),
    demandbase_industry: toStr(row['Demandbase Industry']),
    demandbase_sub_industry: toStr(row['Demandbase Sub Industry']),
    revenue: toInt(row['Revenue']),
    revenue_range: toStr(row['Revenue Range']),
    expected_revenue: toFloat(row['Expected Revenue']),
    arr: toFloat(row['ARR']),
    traffic: toInt(row['Traffic']),

    // Sales assignment
    local_segment: toStr(row['Local Segment']),
    account_region: toStr(row['Account Region']),
    demandbase_account_owner: toStr(row['Demandbase - Account Owner Name']),
    sales_region: toStr(row['Sales Region']),
    sales_sub_region: toStr(row['Sales Sub-Region']),
    sales_segment: toStr(row['Sales Segment']),

    // Technology flags
    has_bigcommerce: toBool(row['Bigcommerce Technology']),
    has_commercetools: toBool(row['Commercetools Technology']),
    has_magento: toBool(row['Magento Technology']),
    has_magento_open_source: toBool(row['Magento Open Source Technology']),
    has_salesforce_b2b_commerce: toBool(row['Salesforce B2b Commerce Technology']),
    has_salesforce_commerce_cloud: toBool(row['Salesforce Commerce Cloud Technology']),
    has_shopify_hosted: toBool(row['Shopify Hosted Technology']),
    has_shopify_plus: toBool(row['Shopify Plus Technology']),
    has_shopify: toBool(row['Shopify Technology']),
    has_spryker: toBool(row['Spryker Technology']),

    // ZoomInfo company data
    zi_company_name: toStr(row['Company Name']),
    zi_company_description: toStr(row['Company Description']),
    zi_website: toStr(row['Website']),
    zi_founded_year: toInt(row['Founded Year']),
    zi_company_hq_phone: toStr(row['Company HQ Phone']),
    zi_fax: toStr(row['Fax']),
    zi_ticker: toStr(row['Ticker']),

    // ZoomInfo financials
    zi_revenue_thousands: toInt(row['Revenue (in 000s USD)']),
    zi_revenue_range: toStr(row['Revenue Range (in USD)']),
    zi_est_marketing_budget_thousands: toInt(row['Est. Marketing Department Budget (in 000s USD)']),
    zi_est_finance_budget_thousands: toInt(row['Est. Finance Department Budget (in 000s USD)']),
    zi_est_it_budget_thousands: toInt(row['Est. IT Department Budget (in 000s USD)']),
    zi_est_hr_budget_thousands: toInt(row['Est. HR Department Budget (in 000s USD)']),

    // ZoomInfo employees
    zi_employees: toInt(row['Employees']),
    zi_employee_range: toStr(row['Employee Range']),
    zi_employee_growth_1yr: toFloat(row['Past 1 Year Employee Growth Rate']),
    zi_employee_growth_2yr: toFloat(row['Past 2 Year Employee Growth Rate']),

    // ZoomInfo industry
    zi_sic_code_1: toStr(row['SIC Code 1']),
    zi_sic_code_2: toStr(row['SIC Code 2']),
    zi_sic_codes: toStr(row['SIC Codes']),
    zi_naics_code_1: toStr(row['NAICS Code 1']),
    zi_naics_code_2: toStr(row['NAICS Code 2']),
    zi_naics_codes: toStr(row['NAICS Codes']),
    zi_primary_industry: toStr(row['Primary Industry']),
    zi_primary_sub_industry: toStr(row['Primary Sub-Industry']),
    zi_all_industries: toStr(row['All Industries']),
    zi_all_sub_industries: toStr(row['All Sub-Industries']),
    zi_industry_hierarchical_category: toStr(row['Industry Hierarchical Category']),
    zi_secondary_industry_hierarchical_category: toStr(row['Secondary Industry Hierarchical Category']),

    // ZoomInfo digital presence
    zi_alexa_rank: toInt(row['Alexa Rank']),
    zi_profile_url: toStr(row['ZoomInfo Company Profile URL']),
    zi_linkedin_url: toStr(row['LinkedIn Company Profile URL']),
    zi_facebook_url: toStr(row['Facebook Company Profile URL']),
    zi_twitter_url: toStr(row['Twitter Company Profile URL']),

    // ZoomInfo corporate
    zi_ownership_type: toStr(row['Ownership Type']),
    zi_business_model: toStr(row['Business Model']),
    zi_certified_active: toStr(row['Certified Active Company']),
    zi_certification_date: toStr(row['Certification Date']),
    zi_defunct_company: toStr(row['Defunct Company']),
    zi_company_is_acquired: toStr(row['Company Is Acquired']),
    zi_number_of_locations: toInt(row['Number of Locations']),

    // ZoomInfo funding
    zi_total_funding_thousands: toInt(row['Total Funding Amount (in 000s USD)']),
    zi_recent_funding_thousands: toInt(row['Recent Funding Amount (in 000s USD)']),
    zi_recent_funding_round: toStr(row['Recent Funding Round']),
    zi_recent_funding_date: toStr(row['Recent Funding Date']),
    zi_recent_investors: toStr(row['Recent Investors']),
    zi_all_investors: toStr(row['All Investors']),

    // ZoomInfo address
    zi_street_address: toStr(row['Company Street Address']),
    zi_city: toStr(row['Company City']),
    zi_state: toStr(row['Company State']),
    zi_zip_code: toStr(row['Company Zip Code']),
    zi_country: toStr(row['Company Country']),
    zi_full_address: toStr(row['Full Address']),

    // ZoomInfo corporate hierarchy
    zi_ultimate_parent_id: toStr(row['Company ID (Ultimate Parent)']),
    zi_ultimate_parent_name: toStr(row['Entity Name (Ultimate Parent)']),
    zi_immediate_parent_id: toStr(row['Company ID (Immediate Parent)']),
    zi_immediate_parent_name: toStr(row['Entity Name (Immediate Parent)']),
    zi_immediate_parent_relationship: toStr(row['Relationship (Immediate Parent)']),

    // ZoomInfo match quality
    zi_match_status: toStr(row['Match status']),
    zi_match_insight_name: toStr(row['Company Name Match Insight']),
    zi_match_insight_domain: toStr(row['Company Domain Match Insight']),
    zi_match_insight_address: toStr(row['Company Address Match Insight']),
    zi_match_insight_phone: toStr(row['Company Phone Match Insight']),
    zi_match_insight_id: toStr(row['Company ID Match Insight']),
    zi_match_insight_social: toStr(row['Company Social URL Match Insight']),

    // Metadata
    source_file: SOURCE_FILE,
  };
}

// =============================================================================
// MAIN
// =============================================================================

async function main() {
  console.log('='.repeat(60));
  console.log('WHALE COMPOSITE IMPORT');
  console.log('='.repeat(60));

  // Read CSV
  console.log(`\nReading CSV: ${CSV_PATH}`);
  const csvContent = fs.readFileSync(CSV_PATH, 'utf-8');

  const records = parse(csvContent, {
    columns: true,
    skip_empty_lines: true,
    trim: true,
  });

  console.log(`Parsed ${records.length} records from CSV`);

  // Map rows
  const mappedRows = records
    .map((row: Record<string, string>) => mapRow(row))
    .filter((row: Record<string, unknown>) => row.domain); // Skip empty rows

  console.log(`Mapped ${mappedRows.length} valid records`);

  // Batch insert (Supabase has limits, use batches of 100)
  const BATCH_SIZE = 100;
  let inserted = 0;
  let errors = 0;

  for (let i = 0; i < mappedRows.length; i += BATCH_SIZE) {
    const batch = mappedRows.slice(i, i + BATCH_SIZE);

    const { data, error } = await supabase
      .from('whale_composite')
      .upsert(batch, { onConflict: 'domain' });

    if (error) {
      console.error(`Batch ${i / BATCH_SIZE + 1} error:`, error.message);
      errors += batch.length;
    } else {
      inserted += batch.length;
      process.stdout.write(`\rInserted: ${inserted}/${mappedRows.length}`);
    }
  }

  console.log('\n');
  console.log('='.repeat(60));
  console.log('IMPORT COMPLETE');
  console.log('='.repeat(60));
  console.log(`Total records: ${mappedRows.length}`);
  console.log(`Inserted/Updated: ${inserted}`);
  console.log(`Errors: ${errors}`);
}

main().catch(console.error);
